controllers/config/k3osconfig_controller.go

```diff
@@ -22,151 +22,163 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-func (r *K3OSConfigReconciler) SetupWithManager(mgr ctrl.Manager, options ...Option) error {
-	r.defaultRequeueResponse = ctrl.Result{RequeueAfter: time.Second * 30}
+	// 4. sync node labels
+	labeler := nodes.NewLabeler()
+	if config.Spec.SyncNodeLabels {
+		if err = labeler.Reconcile(node, nodeConfig.K3OS.Labels); err == nil {
+			updateNode = true
+		} else if err = resultError(err); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
 
-	for _, option := range options {
-		if _, ok := option.(*requireLeaderElectionOpt); ok {
-			r.leader = true
+	// 5. sync node taints
+	tainter := nodes.NewTainter()
+	if config.Spec.SyncNodeTaints {
+		if err = tainter.Reconcile(node, nodeConfig.K3OS.Taints); err == nil {
+			updateNode = true
+		} else if err = resultError(err); err != nil {
+			return ctrl.Result{}, err
 		}
 	}
 
-	if !r.leader { // wrap manager so this can run without a leader lease
-		mgr = &nonLeaderLeaseNeedingManagerWrapper{Manager: mgr}
+	if updateNode {
+		if err = r.updateNode(ctx, node); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.logger.Info("updated node", "labels", node.GetLabels(), "updatedLabels", labeler.GetUpdatedLabels(), "taints", node.Spec.Taints)
 	}
 
-	// cannot inject via inject.LoggerInto because `leader` field isn't set at that point
-	r.logger = mgr.GetLogger().
-		WithName("controllers").
-		WithName("K3OSConfig").
-		WithValues("podName", os.Getenv("HOSTNAME"), "leader", r.leader)
+	return ctrl.Result{}, nil
+}
 
-	return ctrl.NewControllerManagedBy(mgr).For(&configv1alpha1.K3OSConfig{}).Complete(r)
+func (r *K3OSConfigReconciler) getNodeConfig(ctx context.Context, nodeName string) (*nodes.Config, error) {
+	secret, err := r.clientset.CoreV1().Secrets(r.namespace).Get(ctx, consts.NodeConfigSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	nodeConfigBytes, ok := secret.Data[nodeName]
+	if !ok {
+		err = fmt.Errorf("failed to find node %q in config (keys: %v)", nodeName, secretKeys(secret))
+		return nil, err
+	}
+	return nodes.ParseConfig(nodeConfigBytes)
 }
 
-// Interface implementations for dependency injection
-var _ inject.Client = (*K3OSConfigReconciler)(nil)
-var _ inject.Scheme = (*K3OSConfigReconciler)(nil)
+func (r *K3OSConfigReconciler) getNode(ctx context.Context, nodeName string) (*corev1.Node, error) {
+	node, err := r.nodeLister.Get(nodeName)
+	if err != nil {
+		return nil, err
+	}
+	return node.DeepCopy(), nil
+}
 
-// InjectClient satisfies the inject.Client interface.
-func (r *K3OSConfigReconciler) InjectClient(client client.Client) error {
-	r.client = client
+func (r *K3OSConfigReconciler) updateNode(ctx context.Context, node *corev1.Node) error {
+	// TODO: switch to a better way that either implements retries or switch to patching
+	// the node (which would be faster anyways). The only fields we need to update are
+	// the labels, the taints and the annotations (for state tracking). 🤔
```

Comment:
Drop the silly emoji. One emoji in the README is enough for this project.

```suggestion
	// the labels, the taints and the annotations (for state tracking).
```

Reply:
done